ocean-client/src/vault-maxi.ts

```diff
@@ -76,34 +73,50 @@ export async function main(event: maxiEvent): Promise<Object> {
-        return {
-            statusCode: result ? 200 : 500
+    if (settings.stateInformation && settings.stateInformation.state !== ProgramState.Waiting) {
+        const information = settings.stateInformation
+        console.log("last execution stopped state " + information.state)
+        console.log(" for tx " + information.tx)
+        console.log(" on block height " + information.blockHeight)
+
+        let shouldCleanUp = false
+
+        // 2022-03-09 Krysh: input of kuegi
+        // if we are on state waiting for last transaction and we are still on same block height since last execution
+        // then we should wait for next block
+        const currentBlockHeight = await program.getBlockHeight()
+        if (information.state === ProgramState.WaitingForLastTransaction && information.blockHeight === currentBlockHeight) {
+            console.log("waiting for next block")
+            await program.waitForBlockAfter(currentBlockHeight)
+        }
+        // 2022-03-09 Krysh: only clean up if it is really needed, otherwise we are fine and can proceed like normally
+        if (VaultMaxiProgram.shouldCleanUpBasedOn(information.tx as VaultMaxiProgramTransaction)) {
+            console.log("need to clean up")
+            result = await program.cleanUp(vault, telegram)
+            const cleanUpVaultCheck = await program.getVault() as LoanVaultActive
+            await telegram.log("executed clean-up part of script " + (result ? "successfull" : "with problems") + ". vault ratio after clean-up " + cleanUpVaultCheck.collateralRatio)
+
+            await program.updateToState(result ? ProgramState.Waiting : ProgramState.Error, VaultMaxiProgramTransaction.None)
+            return {
+                statusCode: result ? 200 : 500
+            }
         }
     }
 
-    await store.updateToState(VaultMaxiState.Start)
     const collateralRatio = Number(vault.collateralRatio)
     console.log("starting with " + collateralRatio + " in vault, target " + settings.minCollateralRatio + " - " + settings.maxCollateralRatio + " token " + settings.LMToken)
 
-    let result = true
     if (0 < collateralRatio && collateralRatio < settings.minCollateralRatio) {
-        await store.updateToState(VaultMaxiState.DecreaseExposure)
         result = await program.decreaseExposure(vault, telegram)
     } else if (collateralRatio < 0 || collateralRatio > settings.maxCollateralRatio) {
-        await store.updateToState(VaultMaxiState.IncreaseExposure)
         result = await program.increaseExposure(vault, telegram)
     }
     vault = await program.getVault() as LoanVaultActive
     await telegram.log("executed script " + (result ? "successfull" : "with problems") + ". vault ratio " + vaultcheck.collateralRatio)
-    await store.updateToState(VaultMaxiState.Finish)
+    await program.updateToState(result ? ProgramState.Waiting : ProgramState.Error, VaultMaxiProgramTransaction.None)
```

Comment:
would move that directly after the if. getVault and telegram might take a bit of time but don't change the state.